Silabs gpio driver

fimohame opened the pull request:
The commit enables the gpio driver for EFR series 2 devices.

fimohame committed:
drivers: gpio: silabs: gpio driver for EFR series 2 devices

Added the gpio driver for EFR series 2 devices.

The SILABS_SISDK_GPIO symbol is added to enable
support for the new GPIO driver.
The SOC_GECKO_GPIO symbol is retained for now to
maintain compatibility with existing drivers and
will be removed in a subsequent commit.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

fimohame commented on dts/arm/silabs/xg24/xg24.dtsi:
```diff
@@ -438,7 +438,7 @@
 
 			gpioa: gpio@5003c030 {
-				compatible = "silabs,gecko-gpio-port";
+				compatible = "silabs,gecko-gpio-port","silabs,gpio-port";
```
Provided #define DT_DRV_COMPAT silabs_gpio_port in the gpio_silabs.c file

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,384 @@
+
+#define NUMBER_OF_PORTS (SIZEOF_FIELD(GPIO_TypeDef, P) / SIZEOF_FIELD(GPIO_TypeDef, P[0]))
+
+struct gpio_silabs_common_config {
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,384 @@
+		return -EIO;
+	}
+
+	IRQ_CONNECT(GPIO_EVEN_IRQn, DT_IRQ_BY_NAME(DT_INST(0, silabs_gpio), gpio_even, priority),
```
addressed

Martinhoff-maker commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,395 @@
+#include <zephyr/irq.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
```
is it really needed ?

fimohame replied:
addressed

Martinhoff-maker commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,395 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <sl_gpio.h>
```
this one is no needed also no ? I think we will only have sl_hal_gpio

fimohame replied:
addressed

Martinhoff-maker commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,395 @@
+
+static int gpio_silabs_pin_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+	sl_gpio_t gpio = {.port = config->gpio_index, .pin = pin};
+	sl_gpio_mode_t mode;
+	bool pin_out;
+	uint16_t out = 0U;
+	sl_status_t status;
```
I suggest reverse christmas tree declaration in general (longest to shortest line length in declaration) (if possible)

fimohame replied:
addressed

Martinhoff-maker commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,395 @@
+			flag = SL_GPIO_INTERRUPT_RISING_EDGE;
+		}
+
+		status = sl_gpio_configure_external_interrupt(&gpio, &int_no, flag, NULL, NULL);
```
why can't you use sl_hal_gpio_configure_external_interrupt ? In general for all function, we supposed to only use sl_hal_gpio function.

Martinhoff-maker commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,395 @@
+	gpio_silabs_common_data_inst.count = 0;
+	sl_status_t status;
+
+	status = sl_gpio_init();
```
It should not ba called IMO, because it touch the clock. We must used clock_control_on(), etc.

Martinhoff-maker reviewed (changes requested):
I haven't gone through all the gpio_silabs.c files but in general:

- We need to only use sl_hal_gpio functions. sl_gpio.h is already a driver in the HAL and it initializes things like clocks which we want to handle in the Zephyr driver.

- We also need to have a new Kconfig symbol in zephyr/modules/hal_silabs/simplicity_sdk like "SILABS_SISDK_GPIO" and make changes to CMakeLists.txt accordingly. You can take inspiration from https://github.com/zephyrproject-rtos/zephyr/pull/92010

- We should update the dtsi files for all Series 2 boards with this new GPIO driver and create an entry in the migration guide 4.3.

- Don't forget to add a binding file for the driver.

Martinhoff-maker reviewed (commented):
As general comment, as done in gpio_silabs_siwx91x.c, it could be good to have clear device name in general : is it the parent device ( the gpio controller) or the port device rather ? than just having "dev" names everywhere.

Adding PM would be perfect also (might be part of another PR)

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,383 @@
+	return 0;
+}
+
+static int gpio_silabs_port_toggle_bits(const struct device *dev, uint32_t mask)
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,383 @@
+			 PRE_KERNEL_1, CONFIG_GPIO_SILABS_COMMON_INIT_PRIORITY,                    \
+			 &gpio_silabs_common_driver_api);
+
+GPIO_CONTROLLER_INIT(0)
```
addressed

fimohame commented on dts/arm/silabs/xg21/xg21.dtsi:
```diff
@@ -347,7 +347,7 @@
 
 			gpioa: gpio@5003c000 {
-				compatible = "silabs,gecko-gpio-port";
+				compatible = "silabs,gecko-gpio-port","silabs,gpio-port";
```
addressed

Martinhoff-maker commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -253,7 +253,15 @@ zephyr_library_sources_ifdef(CONFIG_SOC_GECKO_EMU          ${EMLIB_DIR}/src/em_e
 if(CONFIG_SOC_GECKO_GPIO)
   zephyr_library_sources(
     ${EMLIB_DIR}/src/em_gpio.c
-    ${DRIVER_DIR}/gpio/src/sl_gpio.c
```
I suggest to let this line for compatibility purpose until we removed all the SOC_GECKO symbol from SimplicitySDK CMakeLists

fimohame commented on drivers/gpio/Kconfig.silabs:
```diff
@@ -0,0 +1,19 @@
+
+menuconfig GPIO_SILABS
+	bool "Silabs GPIO driver"
+	default n
```
addressed

rettichschnidi replied:
If it stays n, then it should be removed: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

fimohame replied:
addressed

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+#define GPIO_CONTROLLER_INIT(idx)                                                                  \
+	static void gpio_silabs_irq_connect_##idx(const struct device *dev)                        \
+	{                                                                                          \
+		IRQ_CONNECT(GPIO_EVEN_IRQn,                                                        \
```
Get interrupt number from DTS using `DT_INST_IRQ_BY_NAME(idx, gpio_even, irq)`

fimohame replied:
addressed

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+	static void gpio_silabs_irq_connect_##idx(const struct device *dev)                        \
+	{                                                                                          \
+		IRQ_CONNECT(GPIO_EVEN_IRQn,                                                        \
+			    DT_IRQ_BY_NAME(DT_INST(idx, silabs_gpio), gpio_even, priority),        \
```
Should be `DT_INST_IRQ_BY_NAME(idx, gpio_even, priority)`

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+	{                                                                                          \
+		IRQ_CONNECT(GPIO_EVEN_IRQn,                                                        \
+			    DT_IRQ_BY_NAME(DT_INST(idx, silabs_gpio), gpio_even, priority),        \
+			    gpio_silabs_common_isr, DEVICE_DT_GET(DT_INST(idx, silabs_gpio)), 0);  \
```
`DEVICE_DT_INST_GET(idx)`

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+		IRQ_CONNECT(GPIO_EVEN_IRQn,                                                        \
+			    DT_IRQ_BY_NAME(DT_INST(idx, silabs_gpio), gpio_even, priority),        \
+			    gpio_silabs_common_isr, DEVICE_DT_GET(DT_INST(idx, silabs_gpio)), 0);  \
+		IRQ_CONNECT(GPIO_ODD_IRQn,                                                         \
```
Same comments as for even

fimohame commented on soc/silabs/silabs_s2/xg27/Kconfig:
```diff
@@ -8,9 +8,11 @@ config SOC_SILABS_XG27
 	select CPU_CORTEX_M33
 	select CPU_HAS_ARM_MPU
 	select CPU_HAS_ARM_SAU
+	select CPU_CORTEX_M_HAS_DWT
```
I was getting the below warnings

[101/171] Building C object zephyr/arch/arch/arm/core/mpu/CMakeFiles/arch__arm__core__mpu.dir/arm_mpu.c.obj
/home/fimohame/zephyrproject/zephyr/arch/arm/core/mpu/arm_mpu.c: In function 'z_arm_mpu_init':
/home/fimohame/zephyrproject/zephyr/arch/arm/core/mpu/arm_mpu.c:490:9: note: '#pragma message: Null-Pointer exception detection cannot be configured on un-mapped flash areas'
  490 | #pragma message "Null-Pointer exception detection cannot be configured on un-mapped flash areas"
  
  in order to address this I had added those. 
  
  Basis the discussion I have removed the macro. We need to address this as part of a different PR

fimohame commented on dts/bindings/gpio/silabs,gpio-port.yaml:
```diff
@@ -0,0 +1,23 @@
+  reg:
+    required: true
+
+  peripheral-id:
```
In the dtsi of xg21, xg23, xg24 peripheral-id is present

do you suggest to remove those?

asmellby replied:
Great! For context, peripheral-id is a hack in devicetree to facilitate the driver implementation, and devicetree should in general not have driver-specific content in it -- it should describe the hardware only. It was used by the gecko gpio driver on Series 0 and 1 to get the port number, but you are doing that in a different way in this driver (using the memory address of the port).

fimohame commented on dts/bindings/gpio/silabs,gpio.yaml:
```diff
@@ -0,0 +1,19 @@
+  interrupts:
+    required: true
+
+  location-swo:
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+		}
+
+		sl_hal_gpio_configure_external_interrupt(&gpio, int_no, flag);
+		sl_hal_gpio_enable_interrupts(BIT(pin));
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	for (uint8_t pin = 0; pin < 32; ++pin) {
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+	struct gpio_silabs_common_data *data = dev->data;
+	uint32_t enabled_int, int_status = GPIO->IF;
+
+	for (unsigned int i = 0; int_status && (i < data->count); i++) {
```
addressed

fimohame committed:
soc: silabs: series_s2: Kconfig updates for EFR series 2 devices

Updated Kconfig for EFR series 2 devices.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

fimohame committed:
tests: gpio: boards: add overlay for GPIO test driver

Added the xg29_rb4412a.overlay for GPIO API 1-pin
This overlay sets the configuration
needed to validate the GPIO driver on the
xg29_rb4412a board during test execution.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,376 @@
+		sl_hal_gpio_enable_interrupts(BIT(pin));
+	}
+
+	WRITE_BIT(data->int_enabled_mask, pin, mode != GPIO_INT_DISABLE);
```
Same issue here, `int_no` is the interrupt line that is enabled, not `pin`.

fimohame replied:
addressed

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,379 @@
+
+		if (enabled_int) {
+			GPIO->IF_CLR = enabled_int;
+			gpio_fire_callbacks(&port_data->callbacks, port_dev, enabled_int);
```
No, I don't believe it is fixed.
Consider enabling an interrupt on both PA0 and PB0.
The first interrupt will use interrupt line 0 (EXTIPINSEL0=0), while the second interrupt will use interrupt line 1 (EXTIPINSEL1=0).
When PB0 gets an interrupt, bit 1 in IF will be set, but gpio_fire_callbacks expects bit 0 to be set in the pin mask.
You need to map the interrupt flags onto the corresponding pins before calling the callback function with a mask of _pins_, not a mask of interrupts.

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,412 @@
+	} else {
+		ARRAY_FOR_EACH(data->irq_pin_map, i) {
+			if ((data->int_enabled_mask & BIT(i)) && (data->irq_pin_map[i] == pin)) {
+				sl_hal_gpio_disable_interrupts(BIT(i));
```
What is the purpose of disabling the interrupt in the "enable" branch?
If the purpose is to trick the configure function into reallocating the same interrupt number if an interrupt was already configured to this pin, I would instead just set `int_no = i;` here, since the configure function only tries to allocate if given the magic "unavailable" value.

fimohame replied:
addressed

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,412 @@
+	/* bitmask of pins with interrupt enabled */
+	uint32_t int_enabled_mask;
+	/* maps interrupt line index to GPIO pin number, 0xFF if unused */
+	uint8_t irq_pin_map[16];
```
Could this be added to `struct gpio_silabs_common_data` instead?

I guess a `const struct device *` would need to be added to the port struct to point back to the common node, but that would still be more efficient (`4*4 + 16 = 32` bytes instead of `4*16 = 64` bytes).

It would correspond to the hardware better, since two ports can't disagree on the irq-pin mapping, and it would also set us up better for adding em4wu interrupt support in the future.

fimohame replied:
addressed

asmellby reviewed (commented):
Looks mostly good to me, a few minor comments.

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,416 @@
+		}
+		const struct device *port_dev = data->ports[i];
+		struct gpio_silabs_port_data *port_data = port_dev->data;
+		struct gpio_silabs_common_data *common = port_data->common_dev->data;
```
This can be hoisted outside the for loop -- `dev` is already the gpio peripheral (same as `port_data->common_dev`).

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,416 @@
+static void gpio_silabs_common_isr(const struct device *dev)
+{
+	struct gpio_silabs_common_data *data = dev->data;
+	struct gpio_silabs_common_data *common = data;
```
Unnecessary alias, can just use `data` directly below.

asmellby commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+#endif
+
+#define SILABS_GPIO_PORT_ADDR_SPACE_SIZE sizeof(GPIO_PORT_TypeDef)
+#define GET_SILABS_GPIO_INDEX(node_id)                                                             \
+	(DT_REG_ADDR(node_id) - DT_REG_ADDR(DT_NODELABEL(gpioa))) / SILABS_GPIO_PORT_ADDR_SPACE_SIZE
```
As long as `gpioa` exists (which it does in all series 2 soc .dtsi files), its `status` doesn't matter, the relevant symbols are always emitted in `devicetree_generated.h`. If it's deleted with `/delete-node/` this will fail, though. But the same pattern is used in the gecko GPIO driver, and I haven't heard of any problems with it.

jerome-pouiller replied:
I assume internally, the hal convert the `gpio_index` in `GPIO_PORT_TypeDef *`. It would have helped if we would hal would have accepted  `GPIO_PORT_TypeDef *` argument.

jerome-pouiller commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+
+#if GPIO_SILABS_COMMON_INIT_PRIORITY >= CONFIG_GPIO_INIT_PRIORITY
+#error GPIO_SILABS_COMMON_INIT_PRIORITY must be less than CONFIG_GPIO_INIT_PRIORITY
+#endif
```
Since the GPIO_Ports are children of GPIO_Common, I believe Zephyr would raise an error, even without this check.

fimohame replied:
addressed

jerome-pouiller commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+};
+
+static inline void gpio_silabs_add_port(struct gpio_silabs_common_data *data,
+					const struct device *dev)
```
Indentation

fimohame replied:
addressed

jerome-pouiller commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		ARRAY_FOR_EACH(common->irq_pin_map, i) {
+			if ((data->int_enabled_mask & BIT(i)) && (common->irq_pin_map[i] == pin)) {
```
Maybe `irq_pin_map` could be replaced by calls to  `sl_hal_gpio_get_external_interrupt_number()` (this is prefered solution because we prefer a single source of trust).

jerome-pouiller commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	uint32_t int_status = GPIO->IF;
+
+	ARRAY_FOR_EACH(data->ports, i) {
+		if (!int_status) {
```
This early exit make this function more complex to understand and it probably does not improve the performance.

fimohame replied:
addressed

jerome-pouiller commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	.port_clear_bits_raw = gpio_silabs_port_clear_bits_raw,
+	.port_toggle_bits = gpio_silabs_port_toggle_bits,
+	.pin_interrupt_configure = gpio_silabs_pin_interrupt_configure,
+	.manage_callback = gpio_silabs_port_manage_callback,
```
nit: you can align the `=` signs.

fimohame replied:
addressed

jhedberg commented:
@fimohame looks like there were plenty of valid points raised by @jerome-pouiller which came a bit too late. Will you work on a follow-up PR to address those?

fimohame commented:
@jerome-pouiller @jhedberg @asmellby @Martinhoff-maker

I have addressed the comments in PR: https://github.com/zephyrproject-rtos/zephyr/pull/94251

Could you please review the changes?

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+		.clock = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(idx)),                                  \
+		.clock_cfg = SILABS_DT_INST_CLOCK_CFG(idx),                                        \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(idx, gpio_silabs_common_init, NULL, &gpio_silabs_common_data_inst,   \
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+		return 0;                                                                          \
+	}                                                                                          \
+	static const struct gpio_silabs_port_config gpio_silabs_port_##n##_config = {              \
+		.common = {.port_pin_mask = (gpio_port_pins_t)(-1)},                               \
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	static int gpio_silabs_port_##n##_init(const struct device *dev)                           \
+	{                                                                                          \
+		struct gpio_silabs_port_data *data = dev->data;                                    \
+		data->common_dev = DEVICE_DT_GET(DT_PARENT(n));                                    \
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	};                                                                                         \
+	static struct gpio_silabs_port_data gpio_silabs_port_##n##_data;                           \
+	DEVICE_DT_DEFINE(n, gpio_silabs_port_##n##_init, NULL, &gpio_silabs_port_##n##_data,       \
+			 &gpio_silabs_port_##n##_config, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,  \
```
addressed

fimohame commented on drivers/gpio/Kconfig.silabs:
```diff
@@ -0,0 +1,19 @@
+
+config GPIO_SILABS_COMMON_INIT_PRIORITY
+	int "Common initialization priority"
+	depends on GPIO_SILABS
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+
+	ARRAY_FOR_EACH(data->irq_pin_map, i) {
+		data->irq_pin_map[i] = 0xFF;
+	}
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	{                                                                                          \
+		struct gpio_silabs_port_data *data = dev->data;                                    \
+		data->common_dev = DEVICE_DT_GET(DT_PARENT(n));                                    \
+		gpio_silabs_add_port(&gpio_silabs_common_data_inst, dev);                          \
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+};
+
+static DEVICE_API(gpio, gpio_common_driver_api) = {
+	.manage_callback = gpio_silabs_port_manage_callback,
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+		}
+		const struct device *port_dev = data->ports[i];
+		struct gpio_silabs_port_data *port_data = port_dev->data;
+		uint32_t enabled_int = int_status & port_data->int_enabled_mask;
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+			mode = SL_GPIO_MODE_INPUT_PULL;
+			out = 1U;
+		} else if (flags & GPIO_PULL_DOWN) {
+			mode = SL_GPIO_MODE_INPUT_PULL;
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+	} else if (flags & GPIO_INPUT) {
+		if (flags & GPIO_PULL_UP) {
+			mode = SL_GPIO_MODE_INPUT_PULL;
+			out = 1U;
```
addressed

fimohame commented on drivers/gpio/gpio_silabs.c:
```diff
@@ -0,0 +1,415 @@
+
+#ifdef CONFIG_GPIO_GET_CONFIG
+static int gpio_silabs_pin_get_config(const struct device *dev, gpio_pin_t pin,
+				      gpio_flags_t *out_flags)
```
addressed
